tests/capture/test_base_interpreter.py

```diff
@@ -0,0 +1,439 @@
+        # first eqn mul, second RY
+        assert inner_jaxpr.eqns[1].primitive == qml.RY._primitive
+        assert len(inner_jaxpr.eqns) == 2
+
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 2.0, 1)
+
+        qml.assert_equal(q.queue[0], qml.ctrl(qml.RY(jax.numpy.array(6.0), 0), 1))
+
+    def test_cond(self):
+        """Test the cond higher order primitive."""
+
+        @SimplifyInterpreter()
+        def f(x, control):
+
+            def true_fn(y):
+                _ = qml.RY(y, 0) ** 2
+
+            def false_fn(y):
+                _ = qml.adjoint(qml.RX(y, 0))
+
+            qml.cond(control, true_fn, false_fn)(x)
+
+        jaxpr = jax.make_jaxpr(f)(0.5, False)
+        assert jaxpr.eqns[0].primitive == cond_prim
+
+        branch1 = jaxpr.eqns[0].params["jaxpr_branches"][0]
+        assert len(branch1.eqns) == 2
+        assert branch1.eqns[1].primitive == qml.RY._primitive
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(branch1, [], 0.5)
+        qml.assert_equal(q.queue[0], qml.RY(2 * jax.numpy.array(0.5), 0))
+
+        branch2 = jaxpr.eqns[0].params["jaxpr_branches"][1]
+        assert len(branch2.eqns) == 2
+        assert branch2.eqns[1].primitive == qml.RX._primitive
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(branch2, [], 0.5)
+        qml.assert_equal(q.queue[0], qml.RX(jax.numpy.array(-0.5), 0))
+
+        assert jaxpr.eqns[0].params["n_args"] == 1
+        assert jaxpr.eqns[0].params["n_consts_per_branch"] == [0, 0]
+
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 2.4, True)
+
+        qml.assert_equal(q.queue[0], qml.RY(jax.numpy.array(4.8), 0))
+
+        with qml.queuing.AnnotatedQueue() as q:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 1.23, False)
+
+        qml.assert_equal(q.queue[0], qml.RX(jax.numpy.array(-1.23), 0))
+
+    def test_cond_no_false_branch(self):
+        """Test transforming a cond HOP when no false branch exists."""
+
+        @SimplifyInterpreter()
+        def f(control):
+
+            @qml.cond(control)
+            def f():
+                _ = qml.X(0) @ qml.X(0)
+
+            f()
+
+        jaxpr = jax.make_jaxpr(f)(True)
+
+        assert jaxpr.eqns[0].params["jaxpr_branches"][-1] is None  # no false branch
+
+        with qml.queuing.AnnotatedQueue() as q_true:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, True)
+
+        qml.assert_equal(q_true.queue[0], qml.I(0))
+
+        with qml.queuing.AnnotatedQueue() as q_false:
+            jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, False)
+
+        assert len(q_false.queue) == 0
+
+    def test_for_loop(self):
+        """Test the higher order for loop registration."""
+
+        @SimplifyInterpreter()
+        def f(n):
+
+            @qml.for_loop(n)
+            def g(i):
+                qml.adjoint(qml.X(i))
```

Comment:
Having `qml.adjoint` here seems pointless. If the intention is to also test that the interpreter is able to handle nested higher-order primitives elegantly, then the `X` should be replaced by an operator that isn't self-adjoint.

Reply:
So `qml.adjoint(qml.RX(x, 0))` instead maybe? The point is to check that the simplification transform is called on the jaxpr inside the for loop.

Reply:
using `qml.adjoitn(qml.X(i))` since it's easy enough to check that the new jaxpr should just have an `X` gate, instead of an adjoint of an x gate.